src/globus_sdk/experimental/globus_app/validating_token_storage/validators.py

```diff
@@ -0,0 +1,289 @@
+            return
+
+        if context.token_data_identity_id != context.prior_identity_id:
+            raise IdentityMismatchError(
+                "Detected a change in identity associated with the token data.",
+                stored_id=context.prior_identity_id,
+                new_id=context.token_data_identity_id,
+            )
+
+
+class ScopeRequirementsValidator(TokenDataValidator):
+    def __init__(
+        self,
+        scope_requirements: t.Mapping[str, t.Sequence[globus_sdk.Scope]],
+        consent_client: globus_sdk.AuthClient,
+    ) -> None:
+        self.scope_requirements = scope_requirements
+        self.consent_client: globus_sdk.AuthClient = consent_client
+        self._cached_consent_forest: ConsentForest | None = None
+
+    def before_store(
+        self,
+        token_data_by_resource_server: t.Mapping[str, TokenStorageData],
+        context: TokenValidationContext,
+    ) -> None:
+        """
+        Validate the token data against scope requirements, but do not check
+        dependent scopes before storage.
+
+        :param token_data_by_resource_server: The data to validate.
+        :param context: The validation context object, containing state of the system at
+            the time of validation.
+        """
+        identity_id = context.token_data_identity_id or context.prior_identity_id
+        for resource_server, token_data in token_data_by_resource_server.items():
+            self._validate_token_data_meets_scope_requirements(
+                resource_server=resource_server,
+                token_data=token_data,
+                identity_id=identity_id,
+                eval_dependent=False,
+            )
+
+    def after_retrieve(
+        self,
+        token_data_by_resource_server: t.Mapping[str, TokenStorageData],
+        context: TokenValidationContext,
+    ) -> None:
+        """
+        Validate the token data against scope requirements, including dependent
+        scope requirements.
+
+        :param token_data_by_resource_server: The data to validate.
+        :param context: The validation context object, containing state of the system at
+            the time of validation.
+        """
+        identity_id = context.token_data_identity_id or context.prior_identity_id
+        for token_data in token_data_by_resource_server.values():
+            self._validate_token_data_meets_scope_requirements(
+                resource_server=token_data.resource_server,
+                token_data=token_data,
+                identity_id=identity_id,
+            )
+
+    def _validate_token_data_meets_scope_requirements(
+        self,
+        *,
+        resource_server: str,
+        token_data: TokenStorageData,
+        identity_id: str | None,
+        eval_dependent: bool = True,
+    ) -> None:
+        """
+        Given a particular resource server/token data, evaluate whether the token +
+            user's consent forest meet the attached scope requirements.
```

Comment:
This dangling "+" catches my eye as a shortcut for writing but not necessarily comprehension.

Reply:
Yeah, this is lift and shift of an existing piece of code, of course. I'll adjust this one, but I do try to keep such changes separate from the functional rework. A dedicated commit for the various doc fixes will do the trick.